src/D2L.Bmx/UpdateHandler.cs

```diff
@@ -0,0 +1,130 @@
+using System.Formats.Tar;
+using System.IO.Compression;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace D2L.Bmx;
+
+internal class UpdateHandler {
+
+	public async Task HandleAsync() {
+		using var httpClient = new HttpClient();
+		var releaseData = await GithubUtilities.GetLatestReleaseDataAsync();
+		var localVersion = Assembly.GetExecutingAssembly().GetName().Version;
+		var latestVersion = new Version( GithubUtilities.GetReleaseVersion( releaseData ) );
+		if( latestVersion <= localVersion ) {
+			Console.WriteLine( $"Already own the latest version {latestVersion}" );
+			return;
+		}
+		string archiveName = GetOSFileName();
+		string downloadUrl = releaseData?.Assets?.FirstOrDefault( a => a.Name == archiveName )?.BrowserDownloadUrl
+			?? string.Empty;
+		if( string.IsNullOrWhiteSpace( downloadUrl ) ) {
+			throw new BmxException( "Failed to get update download url" );
+		}
+
+		string downloadPath = Path.GetTempFileName();
+
+		string? currentFilePath = Environment.ProcessPath;
+		if( string.IsNullOrEmpty( currentFilePath ) ) {
+			throw new BmxException( "BMX could not update" );
+		}
+		string currentDirectory = Path.GetDirectoryName( currentFilePath )!;
+		string backupPath = $"{BmxPaths.OLD_BMX_VERSION_FILE_NAME}v{localVersion}.old.bak";
+
+		try {
+			File.Move( currentFilePath, backupPath );
+		} catch( IOException ex ) {
+			throw new BmxException( "Could move the old version, try with elevated permissions", ex );
+		} catch {
+			throw new BmxException( "BMX could not update" );
+		}
+
+		try {
+			var archiveRes = await httpClient.GetAsync( downloadUrl, HttpCompletionOption.ResponseHeadersRead );
+			using( var fs = new FileStream( downloadPath, FileMode.Create, FileAccess.Write, FileShare.None ) ) {
+				await archiveRes.Content.CopyToAsync( fs );
+				await fs.FlushAsync();
+				fs.Dispose();
+			}
```

Comment:
* don't call dispose when you're already in scope of a using
* prefer [using declaration](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/using#using-declaration) to reduce indentation wherever possible. same here: https://github.com/Brightspace/bmx/pull/418/files#diff-48acde0deb16dbeda7f36eb7087c301cf5a486a4fe2044fc77522fdd67191353R121

```suggestion
			using var fs = new FileStream( downloadPath, FileMode.Create, FileAccess.Write, FileShare.None );
			await archiveRes.Content.CopyToAsync( fs );
			await fs.FlushAsync();
```